network/p2p/libp2pNode.go

```diff
@@ -81,66 +91,62 @@ type Node struct {
 	connMgr              *ConnManager
 }
 
-func NewLibP2PNode(logger zerolog.Logger,
-	id flow.Identifier,
-	address string,
-	conMgr *ConnManager,
-	key fcrypto.PrivateKey,
-	allowList bool,
-	rootBlockID string,
-	pingInfoProvider PingInfoProvider,
-	psOption ...pubsub.Option) (*Node, error) {
+type NodeOption func(node *Node) error
 
-	libp2pKey, err := PrivKey(key)
-	if err != nil {
-		return nil, fmt.Errorf("could not generate libp2p key: %w", err)
+func WithLibP2PHost(host host.Host) NodeOption {
+	return func(node *Node) error {
+		node.host = host
+		return nil
 	}
+}
 
-	flowLibP2PProtocolID := generateFlowProtocolID(rootBlockID)
-
-	ctx, cancel := context.WithCancel(context.Background())
-
-	libP2PHost, connGater, pubSub, err := bootstrapLibP2PHost(ctx,
-		logger,
-		address,
-		conMgr,
-		libp2pKey,
-		allowList,
-		psOption...)
+func WithConnectionManager(conMgr *ConnManager) NodeOption {
+	return func(node *Node) error {
+		node.connMgr = conMgr
+		return nil
+	}
+}
 
-	if err != nil {
-		cancel()
-		return nil, fmt.Errorf("could not bootstrap libp2p host: %w", err)
+func WithPubSub(pubsub *pubsub.PubSub) NodeOption {
+	return func(node *Node) error {
+		node.pubSub = pubsub
+		return nil
 	}
+}
 
-	pingLibP2PProtocolID := generatePingProtcolID(rootBlockID)
-	pingService := NewPingService(libP2PHost, pingLibP2PProtocolID, pingInfoProvider, logger)
+func NewLibP2PNode(id flow.Identifier, rootBlockID string, logger zerolog.Logger, options ...NodeOption) (*Node, error) {
 
-	n := &Node{
-		connGater:            connGater,
-		host:                 libP2PHost,
-		pubSub:               pubSub,
+	ctx, cancel := context.WithCancel(context.Background())
+
+	node := &Node{
+		id:                   id,
+		flowLibP2PProtocolID: generateFlowProtocolID(rootBlockID),
+		ctx:                  ctx,
 		cancel:               cancel,
 		logger:               logger,
 		topics:               make(map[flownet.Topic]*pubsub.Topic),
 		subs:                 make(map[flownet.Topic]*pubsub.Subscription),
-		id:                   id,
-		flowLibP2PProtocolID: flowLibP2PProtocolID,
-		pingService:          pingService,
-		connMgr:              conMgr,
 	}
 
-	ip, port, err := n.GetIPPort()
+	for _, opt := range options {
+		err := opt(node)
+		if err != nil {
```

Comment:
Imo, we should remove the error return from the type definition of `NodeOption`. Conceptally, functional options should just be a mechanism for passing in configurations to a constructor, and errors shouldn't occur as a result of configuring an object.

It is a different story if errors occur when trying to create subcomponents in the constructor itself.

I'll add more thoughts on how we could refactor below.